File: python/src/aoc2025/scaffold.py
"""Scaffolding for new day solutions.

Creates the necessary files for a new day's solution.
"""

from pathlib import Path

# Paths relative to this file
DAYS_DIR = Path(__file__).parent / "days"
TESTS_DIR = Path(__file__).parent.parent.parent / "tests"
DATA_DIR = Path(__file__).parent.parent.parent.parent.parent / "data"


def scaffold_day(day: int) -> None:
    """Scaffold a new day's solution files."""
    day_str = f"{day:02d}"

    # Create the day module file
    day_path = DAYS_DIR / f"day{day_str}.py"
    if day_path.exists():
        raise FileExistsError(f"Day {day} module already exists at {day_path}")

    day_template = _generate_day_template(day)
    day_path.write_text(day_template)
    print(f"  📄 Created: {day_path}")

    # Create the test file
    test_path = TESTS_DIR / f"test_day{day_str}.py"
    if not test_path.exists():
        test_template = _generate_test_template(day)
        test_path.write_text(test_template)
        print(f"  📄 Created: {test_path}")

    # Create the example input file
    example_path = DATA_DIR / "examples" / f"{day_str}.txt"
    if not example_path.exists():
        example_path.parent.mkdir(parents=True, exist_ok=True)
        example_path.write_text("# Paste example input here\n")
        print(f"  📄 Created: {example_path}")

    # Print instructions for manual step
    print()
    print("📝 Manual step required:")
    print(f"   Add the import to src/aoc2025/days/__init__.py:")
    print()
    print(f"   from . import day{day_str}")
    print()
    print(f"   And add to the DAYS dict:")
    print()
    print(f"   {day}: day{day_str},")


def _generate_day_template(day: int) -> str:
    """Generate the template for a new day module."""
    return f'''"""Day {day:02d}: [Title]

https://adventofcode.com/2025/day/{day}
"""


def part_one(input_text: str) -> int | None:
    """Solve part one."""
    lines = input_text.strip().split("\\n")

    # TODO: Implement solution
    return None


def part_two(input_text: str) -> int | None:
    """Solve part two."""
    lines = input_text.strip().split("\\n")

    # TODO: Implement solution
    return None
'''


def _generate_test_template(day: int) -> str:
    """Generate the template for a new test file."""
    day_str = f"{day:02d}"
    return f'''"""Tests for Day {day:02d}."""

import pytest

from aoc2025.days import day{day_str}


@pytest.fixture
def example_input(load_example):
    """Load the example input for day {day:02d}."""
    return load_example({day})


class TestPart1:
    """Tests for Part 1."""

    def test_example(self, example_input):
        """Test part one with the example input."""
        # TODO: Update expected value from puzzle description
        result = day{day_str}.part_one(example_input)
        assert result is None  # Replace with expected value


class TestPart2:
    """Tests for Part 2."""

    def test_example(self, example_input):
        """Test part two with the example input."""
        # TODO: Update expected value from puzzle description
        result = day{day_str}.part_two(example_input)
        assert result is None  # Replace with expected value
'''

File: rust/src/scaffold.rs
//! Scaffolding for new day solutions
//!
//! Creates the necessary files for a new day's solution.

use std::fs;
use std::path::PathBuf;

/// Scaffold a new day's solution files
pub fn scaffold_day(day: u8) -> Result<(), Box<dyn std::error::Error>> {
    let day_str = format!("{:02}", day);

    // Create the day module file
    let day_path = PathBuf::from(format!("src/days/day{}.rs", day_str));
    if day_path.exists() {
        return Err(format!("Day {} module already exists at {:?}", day, day_path).into());
    }

    let day_template = generate_day_template(day);
    fs::write(&day_path, day_template)?;
    println!("  📄 Created: {:?}", day_path);

    // Create the example input file
    let example_path = PathBuf::from(format!("../data/examples/{}.txt", day_str));
    if !example_path.exists() {
        if let Some(parent) = example_path.parent() {
            fs::create_dir_all(parent)?;
        }
        fs::write(&example_path, "# Paste example input here\n")?;
        println!("  📄 Created: {:?}", example_path);
    }

    // Print instructions for manual step
    println!();
    println!("📝 Manual step required:");
    println!("   Add the following line to src/days/mod.rs:");
    println!();
    println!("   pub mod day{};", day_str);
    println!();
    println!("   And add a match arm in the run_day function:");
    println!();
    println!("   {} => match part {{", day);
    println!("       1 => day{}::part_one(input).map(|v| v as i64),", day_str);
    println!("       2 => day{}::part_two(input).map(|v| v as i64),", day_str);
    println!("       _ => None,");
    println!("   }},");

    Ok(())
}

fn generate_day_template(day: u8) -> String {
    let day_str = format!("{:02}", day);
    format!(
        r#"//! Day {day}: [Title]
//!
//! https://adventofcode.com/2025/day/{day}

/// Solve part one
pub fn part_one(input: &str) -> Option<u64> {{
    let _lines: Vec<&str> = input.lines().collect();
    
    // TODO: Implement solution
    None
}}

/// Solve part two
pub fn part_two(input: &str) -> Option<u64> {{
    let _lines: Vec<&str> = input.lines().collect();
    
    // TODO: Implement solution
    None
}}

#[cfg(test)]
mod tests {{
    use super::*;

    fn example_input() -> String {{
        std::fs::read_to_string("../data/examples/{day_str}.txt")
            .expect("Failed to read example input")
    }}

    #[test]
    fn test_part_one() {{
        let input = example_input();
        // TODO: Update expected value from puzzle description
        assert_eq!(part_one(&input), None);
    }}

    #[test]
    fn test_part_two() {{
        let input = example_input();
        // TODO: Update expected value from puzzle description
        assert_eq!(part_two(&input), None);
    }}
}}
"#,
        day = day,
        day_str = day_str,
    )
}
